src/ContaoCommunityAlliance/DcGeneral/Contao/View/Contao2BackendView/Subscriber/WidgetBuilder.php

```diff
@@ -0,0 +1,465 @@
+    }
+
+    /**
+     * Build a widget for a given property.
+     *
+     * @param PropertyInterface $property    The property.
+     *
+     * @param ModelInterface    $model       The current model.
+     *
+     * @param PropertyValueBag  $inputValues The input values to use (optional).
+     *
+     * @return \Widget
+     *
+     * @SuppressWarnings(PHPMD.Superglobals)
+     */
+    public function buildWidget(
+        PropertyInterface $property,
+        ModelInterface $model,
+        PropertyValueBag $inputValues = null
+    ) {
+        $environment  = $this->getEnvironment();
+        $dispatcher   = $environment->getEventDispatcher();
+        $propertyName = $property->getName();
+        $propExtra    = $property->getExtra();
+        $defName      = $environment->getDataDefinition()->getName();
+        $strClass     = $this->getWidgetClass($property);
+        $varValue     = $this->propertyValueEncoder->decodeValue(
+            $propertyName,
+            $model->getProperty($propertyName),
+            $model
+        );
+
+        if ((isset($propExtra['rgxp']) && in_array($propExtra['rgxp'], array('date', 'time', 'datim')))
+            && empty($propExtra['mandatory'])
+            && is_numeric($varValue) && $varValue == 0
+        ) {
+            /*
+                FIXME TEMPORARY WORKAROUND! To be fixed in the core:
+                @see \Widget::getAttributesFromDca()
+            */
+
+            $varValue = '';
+        }
+
+        // OH: why not $required = $mandatory always? source: DataContainer 226.
+        // OH: the whole prepareForWidget(..) thing is an only mess
+        // Widgets should parse the configuration by themselves, depending on what they need.
+        $propExtra['required'] = ($varValue == '') && !empty($propExtra['mandatory']);
+
+        if ($inputValues) {
+            $values   = new PropertyValueBag($inputValues->getArrayCopy());
+            $newModel = clone $model;
+            $newModel->setId($model->getId());
+            $this->environment->getController()->updateModelFromPropertyBag($newModel, $values);
+            $model = $newModel;
+        }
+
+        $arrConfig = array(
+            'inputType' => $property->getWidgetType(),
+            'label'     => array(
+                $property->getLabel(),
+                $property->getDescription()
+            ),
+            'options'   => $this->getOptionsForWidget($property, $model),
+            'eval'      => $propExtra,
+            // TODO: populate these.
+            // 'foreignKey' => null
+        );
+
+        if (isset($propExtra['reference'])) {
+            $arrConfig['reference'] = $propExtra['reference'];
+        }
+
+        $event = new GetAttributesFromDcaEvent(
+            $arrConfig,
+            $property->getName(),
+            $varValue,
+            $propertyName,
+            $defName,
+            new DcCompat($environment, $model, $propertyName)
+        );
+
+        // FIXME: propagator
```

Comment:
Afaik the propagator was removed from the code. What about the suffixed events, do they still exists or will they get removed as well?